subxt/src/utils/bits.rs

```diff
@@ -0,0 +1,227 @@
+                /// Number of bits that the backing store types holds.
+                const BITS: u32;
+            }
+
+            $(
+                impl $ident for $wrapped {
+                    const FORMAT: StoreFormat = StoreFormat::$ty;
+                    const BITS: u32 = <$wrapped>::BITS;
+                }
+            )*
+        };
+    }
+
+macro_rules! order {
+        ($ident: ident; $($ty: ident),*) => {
+            /// Associates `bitvec::order::BitOrder` trait with corresponding, type-erased `scale_bits::OrderFormat` enum.
+            ///
+            /// Used to decode bit sequences in runtime by providing `scale_bits::OrderFormat` using
+            /// `bitvec`-like type type parameters.
+            pub trait $ident {
+                /// Corresponding `scale_bits::OrderFormat` value.
+                const FORMAT: OrderFormat;
+            }
+
+            $(
+                #[doc = concat!("Type-level value that corresponds to `scale_bits::OrderFormat::", stringify!($ty), "` at run-time")]
+                #[doc = concat!(" and `bitvec::order::BitOrder::", stringify!($ty), "` at the type level.")]
+                #[derive(Clone, Debug, PartialEq, Eq)]
+                pub enum $ty {}
+                impl $ident for $ty {
+                    const FORMAT: OrderFormat = OrderFormat::$ty;
+                }
+            )*
+        };
+    }
+
+store!(BitStore; (U8, u8), (U16, u16), (U32, u32), (U64, u64));
+order!(BitOrder; Lsb0, Msb0);
+
+/// Constructs a run-time format parameters based on the corresponding type-level parameters.
+fn bit_format<Store: BitStore, Order: BitOrder>() -> Format {
+    Format {
+        order: Order::FORMAT,
+        store: Store::FORMAT,
+    }
+}
+
+/// `scale_bits::Bits` generic over the bit store (`u8`/`u16`/`u32`/`u64`) and bit order (LSB, MSB)
+/// used for SCALE encoding/decoding. Uses `scale_bits::Bits`-default `u8` and LSB format underneath.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DecodedBits<Store: BitStore, Order: BitOrder>(
+    Bits,
+    PhantomData<Store>,
+    PhantomData<Order>,
+);
+
+impl<Store: BitStore, Order: BitOrder> DecodedBits<Store, Order> {
+    /// Extracts the underlying `scale_bits::Bits` value.
+    pub fn into_bits(self) -> Bits {
+        self.0
+    }
+
+    /// References the underlying `scale_bits::Bits` value.
+    pub fn to_bits(&self) -> &Bits {
+        &self.0
+    }
+}
+
+impl<Store: BitStore, Order: BitOrder> core::iter::FromIterator<bool>
+    for DecodedBits<Store, Order>
+{
+    fn from_iter<T: IntoIterator<Item = bool>>(iter: T) -> Self {
+        DecodedBits(Bits::from_iter(iter), PhantomData, PhantomData)
+    }
+}
+
+impl<Store: BitStore, Order: BitOrder> codec::Decode for DecodedBits<Store, Order> {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+        /// Equivalent of `BitSlice::MAX_BITS` on 32bit machine.
+        const ARCH32BIT_BITSLICE_MAX_BITS: usize = 0x1fff_ffff;
```

Comment:
I think `bits` below is already u32 and coukd remove some `as` casts I guess

```suggestion
        const ARCH32BIT_BITSLICE_MAX_BITS: u32 = 0x1fff_ffff;
```

Reply:
That's fair, simplified a bit with this turned into `u32`. Thanks!